Import nise

mroderick opened the pull request:
This PR is part of #1453, it replaces fake xhr and fake server with [`nise`](https://github.com/mroderick/nise)

#### To try it out

1. Check out this branch
1. `npm install`
1. `npm test`

mroderick commented:
Added the `MAJOR` label because:

* 944f6cf removes previously deprecated properties
* We might break something

It's better that people only get this release through conscious upgrade, and not part of regular install (damn you loose versioning!!!!)

fearphage commented:
> * 944f6cf removes previously deprecated properties

FYI these changes should end up int he release notes.

mroderick commented:
I have pushed new commits that imports the new nise module from npm.

There are still some deprecations we can remove:

* [`sandbox.stub(obj, 'meth', val)`](https://github.com/sinonjs/sinon/blob/2395f35b01c74eab50a27f38ad903c3d497dd931/lib/sinon/sandbox-stub.js#L13-L18)
* [`sinon.stub(obj, 'meth', fn)`](https://github.com/sinonjs/sinon/blob/2395f35b01c74eab50a27f38ad903c3d497dd931/lib/sinon/stub-descriptor.js#L12-L17)
* [`sinon/util/core`](https://github.com/sinonjs/sinon/blob/2395f35b01c74eab50a27f38ad903c3d497dd931/lib/sinon.js#L7-L26)

Shall we remove them as part of this `MAJOR` release, or save them for another day?

fatso83 commented:
Remove, remove, remove!

fatso83 commented:
I think we can merge this and remove any remaining deprecations in another PR. Anyone opposed?

mroderick commented:
> Anyone opposed?

Me! I am already working on removing the last deprecations, so I'd rather just keep this branch open for a bit longer, than either have `master` unshippable again or having to maintain another legacy branch.

mroderick commented:
OK, I think I have mopped up all the previously deprecated stuff.

If we're going to put out a new `MAJOR` release based on this PR, then I think it deserves a thorough review.

mroderick commented:
I am attending a family event today, so won't be able to contribute anything until tonight at the earliest and perhaps not even until sometime next week.

mroderick commented:
Ping @sinonjs/core do you have any further comments for this PR?

fatso83 commented:
Njet. All good.

fearphage commented on lib/sinon/stub-non-function-property.js:
```diff
@@ -3,15 +3,13 @@
         throw new TypeError("Cannot stub non-existent own property " + valueToString(property));
     }
 
-    object[property] = value;
```
Looks like most of the property tests were removed. I see negative tests, but no positive tests.

mroderick replied:
Stubbing properties is the responsibility of `collection.js`, it has [several tests covering this](https://github.com/sinonjs/sinon/blob/5eca462a5f769c7de1622886be66d76a67f245a1/test/collection-test.js#L131-L182).

fearphage commented on lib/sinon/util/core/index.js:
```diff
@@ -2,7 +2,6 @@
 
 module.exports = {
     calledInOrder: require("./called-in-order"),
-    configureLogError: require("./log_error"),
```
If this was accessible, we should document that it was removed even if it was previously deprecated.

fearphage commented on lib/sinon.js:
```diff
@@ -49,20 +41,13 @@ exports.useFakeTimers = fakeTimers.useFakeTimers;
 exports.timers = fakeTimers.timers;
 
-var event = require("./sinon/util/event");
-exports.Event = deprecated.wrap(event.Event, deprecated.defaultMsg("Event"));
```
Documentation for all of these changes.

fearphage commented on lib/sinon/stub-descriptor.js:
```diff
@@ -1,41 +0,0 @@
-
-// This is deprecated and will be removed in a future version of sinon.
-// We will only consider pull requests that fix serious bugs in the implementation
-function stubDescriptor(object, property, descriptor) {
```
Documentation.

mroderick replied:
This was an internal only file.

That `sinon.stub` only accepts two arguments has been [documented in this very commit](https://github.com/mroderick/sinon/blob/fdea27c9b77ea30e97bd0233853cabb7a77e0b7c/docs/release-source/release/stubs.md#var-stub--sinonstubobject-method-func)

fearphage commented on test/sandbox-test.js:
```diff
@@ -597,6 +597,7 @@ describe("sinonSandbox", function () {
 
         it("allows stubbing setters", function () {
             var object = {
+                foo: undefined,
```
Did this need to be added?

mroderick replied:
It is necessary, otherwise we run into the [input validation of collection.stub, that checks for own properties on stubbing targets](https://github.com/sinonjs/sinon/blob/5eca462a5f769c7de1622886be66d76a67f245a1/lib/sinon/collection.js#L93-L95).

`TypeError: Cannot stub non-existent own property foo`

mroderick commented on lib/sinon/stub-non-function-property.js:
```diff
@@ -3,15 +3,13 @@
 var hasOwnProperty = Object.prototype.hasOwnProperty;
 
-function stubNonFunctionProperty(object, property, value) {
+function stubNonFunctionProperty(object, property) {
```
> This should throw like the other stub method with too many arguments.

It is internal only, and is called inside `sandbox.stub`, which does throw for too many arguments.

mroderick replied:
> The replacement for this usage is not documented

The practice of using `sandbox.stub` for [stubbing out properties has been documented for awhile](http://sinonjs.org/releases/v2.3.8/sandbox/#sandboxstub), admittedly, not very well.

mroderick replied:
I have pushed another commit to improve documentation for stubbing out non-function properties

mroderick commented on lib/sinon/collection.js:
```diff
@@ -85,9 +85,13 @@ var collection = {
+    stub: function stub(object, property) {
         if (arguments.length > 2) {
-            return sandboxStub.apply(this, arguments);
+            throw new TypeError("stub(obj, 'meth', fn) has been removed, see documentation");
```
Indeed, well spotted

mroderick commented:
@fearphage thank you for your thorough review.

I think I have addressed all your comments.

If you're satisfied, then I'll fold the review commit into the appropriate commit and merge it.

fearphage commented on test/stub-test.js:
```diff
@@ -1027,11 +996,7 @@ describe("stub", function () {
-            }});
-
-            assert.equals(obj.prop, 43);
+            var stub = createStub(obj, "prop");
```
This test has no assertions now. Add an assertion or remove the test.

mroderick replied:
That was a mistake! I've fixed this in the offending commit and have rebased my branch

mroderick commented:
I think everything has been addressed. Since #1496 is also a `MAJOR`, I would like to merge that one first and then this one, and then just cut one new `MAJOR` release, instead of two.
